### PR TITLE
impl(GCS+gRPC): new class for bucket names

### PR DESCRIPTION
**Pull request by coryan**
Part of the work for #13910

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/13967)

### CODE REVIEW

File: google/cloud/storage/async/bucket_name.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+/**
+ * Represent bucket names.
+ *
+ * Bucket names must be properly formatted before using the storage gRPC API.
+ *
+ * @note This class makes no effort to validate the bucket id. The service
+ *     imposes some restrictions on [bucket names], such as length and the
+ *     character set. The application must ensure the bucket name meets these
+ *     restrictions, this class performs no validation on its inputs.
```
Comment (devbww) [inline google/cloud/storage/async/bucket_name.h:36]: s/,/ as/

File: google/cloud/storage/async/bucket_name.cc [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+std::ostream& operator<<(std::ostream& os, BucketName const& p) {
+  return os << p.FullName();
+}
+
+StatusOr<BucketName> MakeBucketName(std::string const& full_name) {
+  if (!absl::StartsWith(full_name, kPrefix)) {
+    return internal::InvalidArgumentError(
+        absl::StrCat("missing prefix (", kPrefix,
+                     ") for in bucket name: ", full_name),
```
Comment (devbww) [inline google/cloud/storage/async/bucket_name.cc:51]: s/ for//

File: google/cloud/storage/async/bucket_name.cc [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+bool operator!=(BucketName const& a, BucketName const& b) { return !(a == b); }
+
+std::ostream& operator<<(std::ostream& os, BucketName const& p) {
+  return os << p.FullName();
+}
+
+StatusOr<BucketName> MakeBucketName(std::string const& full_name) {
+  if (!absl::StartsWith(full_name, kPrefix)) {
```
Comment (devbww) [inline google/cloud/storage/async/bucket_name.cc:48]: Consider `absl::ConsumePrefix()`, which does both the "starts with" and the "strip prefix".

**Review by devbww — APPROVED** [review]

File: google/cloud/storage/async/bucket_name.h (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/async/bucket_name.h]: Fixed.

File: google/cloud/storage/async/bucket_name.cc (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/async/bucket_name.cc]: Fixed

File: google/cloud/storage/async/bucket_name.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/async/bucket_name.cc]: Done.

**Pull request merged**